### ISSUE TITLE
Add SingLoRA to PEFT

### ISSUE THREAD

**Issue by DavidBCode** [huggingface/peft#2652]
### Feature request

Hi!

We propose to add SingLoRA to peft. SingLoRA is a novel PEFT method recently published in _SingLoRA: Single‑Matrix Low‑Rank Adaptation_ (https://arxiv.org/pdf/2507.05566).
Instead of LoRA’s two‑matrix adapter BA, SingLoRA learns just one matrix A and the adapter is defined as AA^t. This simpler design cuts resource costs in half and has been shown to boost performance on large language models.
In the original paper, authors show that fine‑tuning LLaMA with SingLoRA outperforms LoRA by 1.1 % in accuracy (89.1% vs 90.2% in GLUE) while using only half the adapter parameters. On the vision side, DreamBooth fine‑tuning of Stable Diffusion 1.5 achieves a 5.4 % improvement in DINO similarity at the same rank with half as many parameters.

Key Benefits
* ~50 % fewer adapter parameters compared to LoRA
* Higher accuracy at the same rank for fundamental tasks such as LLM understanding (+ 1.1 % on GLUE for LLaMA) and image generation DreamBooth on SD 1.5 (+ 5.4 % DINO similarity) with half the parameters
* Single learning rate – no more tuning separate scales for two matrices, leading to more stable training dynamics
* Compatibility with modern LoRA variants such as DoRA

### Your contribution

Open source (non-official) implementation of SingLoRA: https://github.com/kyegomez/SingLoRA/tree/main .
I'm also available to prepare a PR to peft to include SingLoRA upon request :)

**Comment by BenjaminBossan** [comment]
Thanks for proposing to add SingLoRA to PEFT. Are you one of the authors?

I had skimmed the paper some time ago and will just quote what I wrote internally:

> The init and forward pass should be simple to implement in PEFT as a LoRA variant. However, they also have a ramp up schedule for training (as SingLoRA initially is not a no-op, unlike LoRA), which is a bit outside of the PEFT scope.

Do you think that assessment is correct? Note that with LoRA variant, I mean it can be implemented similarly to, for instance, [DoRA](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/variants.py).

**Comment by DavidBCode** [comment]
Thanks for your quick response. Yes, I am the first author.

There is indeed a ramp-up function, but it is really a minor aspect of the paper. The core idea—and the improvements it brings—lies in using a single matrix in LoRA to enhance PEFT. This is unrelated to training schedules or optimizers. The ramp-up function is only meant to ensure that, at initialization, the LoRA fine-tuned model matches the pretrained one. Even a hardcoded constant (e.g., always using t/100, where t is the current step) is sufficient.

I genuinely believe that SingLoRA fits within the scope of the peft package, as it’s a simple variant of LoRA that delivers real-world improvements in both performance and efficiency (in terms of memory and compute). I also think it's easier to implement than DoRA. The only addition needed is a hook to retrieve the current training step. If you'd like, we can work together to figure out how to address that.

Best,
David

**Comment by BenjaminBossan** [comment]
Thanks for explaining this further. I agree that the core implementation should not be too difficult. Regarding the ramp up, I'm sure we can think of something. Good documentation and a nice example (e.g. a vanilla train loop + one example with `Trainer`) should hopefully suffice to instruct users how to apply it correctly.

**Comment by bghira** [comment]
well, it's not ideal, but i noted the custom models topic and the associated docs made it sound easy, so i've created a library to help integrate SingLoRA training with PEFT capable models.

https://github.com/bghira/PEFT-SingLoRA

- it replaces nn.Linear wholesale so probably not the greatest approach, limited in flexibility?
- registers buffers to keep track of the step, but that's possibly not something that would fly for PEFT

**Comment by BenjaminBossan** [comment]
Thanks @bghira for providing this repo, I think it will be helpful when it comes to integrating this in PEFT. I think instead of replacing `nn.Linear`, we can implement it as a LoRA variant in PEFT.

When it comes to registering (non-persistent) buffers for record keeping, I'm not sure if there is not a simpler solution, but in the end I think it can work too. I think the relevant part is here, right?

https://github.com/bghira/PEFT-SingLoRA/blob/42925a127b0d8ea779a548bcf6b4035286a0afc8/src/peft_singlora/layer.py#L175-L177

My issue with that is that we cannot always make the assumption that 1 forward call == 1 training step, think about gradient accumulation. For that reason, it might just be better to require the user to manually make keep track of the progress. There is precedent for that in PEFT with [`update_and_allocate`](https://github.com/huggingface/peft/blob/fa85d10a7f1cdfbe616c232083d6fc8d1c5279c7/src/peft/tuners/adalora/model.py#L320) in AdaLoRA.

**Comment by bghira** [comment]
oh, i didn't even think of that - i hardly ever use gradient accumulation. thanks for the pointer to AdaLoRA I will look at that later today and probably extend/fix the step tracking.

**Comment by bghira** [comment]
i managed to get it mostly working for image training in simpletuner and ran into an issue where bf16 doesn't work with ramp up steps since it makes things a bit unstable. but disabling the ramp up fixes that. the issue currently exists in both public implementations of SingLoRA.

**Comment by BenjaminBossan** [comment]
Great, thanks for running the experiments. I see that you now have `update_singlora_global_step`, I think for the PEFT integration we can add it as a method on the model class itself.

> ran into an issue where bf16 doesn't work with ramp up steps since it makes things a bit unstable

Thanks for the heads up. If it the same with fp16?

**Comment by bghira** [comment]
no, fp16 was fine. i would have added to the model class but as an add-on it didnt need/use one yet.

**Comment by shermansiu** [comment]
Are there any plans to merge https://github.com/bghira/PEFT-SingLoRA as a core feature in `peft`?

**Comment by BenjaminBossan** [comment]
I think the plan is still that @DavidBCode will create a PR to add SingLoRA to PEFT.

**Comment by bghira** [comment]
i've been running extensive tests with it and it majorly underperforms against vanilla LoRA for standard training on image diffusion models (MMDiT) especially when introducing complex tasks, it's just not possible to do with SingLoRA, there isn't enough accuracy?

<img width="2309" height="768" alt="Image" src="https://github.com/user-attachments/assets/ff8ae780-c91f-43de-9e3d-96ea3ff53e99" />

<img width="2309" height="768" alt="Image" src="https://github.com/user-attachments/assets/a5eb71cd-d706-47be-a3c0-e7d744b9aa55" />

<img width="2309" height="768" alt="Image" src="https://github.com/user-attachments/assets/73e3765c-e97e-463d-b1de-93d79d28ce98" />

additionally, although it uses half the parameter count as the original LoRA layers, it doesn't train faster.

**Comment by bghira** [comment]
vanilla LoRA, same dataset, rank, etc

<img width="2686" height="892" alt="Image" src="https://github.com/user-attachments/assets/ba1b2375-e749-4ea6-aa24-48ca782ee38f" />

**Comment by DavidBCode** [comment]
Hi, I'm currently a bit busy and will prepare a PR next week. @bghira You can easily adapt a simple codebase such as  https://github.com/nikhilgsh/loraplus/tree/main/glue to check SingLoRA in a more systematic and quantitative way on LLMs. I will publish examples with the PR.

**Comment by BenjaminBossan** [comment]
> vanilla LoRA, same dataset, rank, etc

With fewer parameters, it is perhaps expected that the learning capacity is lower. Did you try a higher rank for SingLoRA? Of course, if it's too high, the advantages in terms of memory compared to LoRA would also vanish.

**Comment by bghira** [comment]
yes, higher rank is needed but not quite double so there are still wins. but i have no familiarity with LLMs and was interested in exploring adaptation to diffusion domain

**Comment by shermansiu** [comment]
What model, dataset, and rank are you using, specifically? 

And how high of a rank did you use for SingLoRA? Do you need an absurdly high rank to get similar results as LoRA (worst-case, this would be the full rank), or does that not work either?

**Comment by bghira** [comment]
@shermansiu the experiment used LibreFLUX.1 as a base (a de-distilled 12B Schnell model) and up to rank 512, didn't have the resources to try full-rank (but will go further if i receive access to that)

used the 1.2M OmniEdit dataset.

i'm assuming LLMs respond more favourably to this formulation, or their deficits just aren't readily apparent since there's not as much visual inspection that can be done with text outputs unlike video or image models.

**Comment by shermansiu** [comment]
Thanks! This information is quite useful.

**Comment by bghira** [comment]
this was completed?

**Comment by BenjaminBossan** [comment]
@DavidBCode Do you still plan to work on this?

**Comment by DavidBCode** [comment]
@BenjaminBossan Sorry, I’ve been busier than planned. I’ll do it soon and keep you updated. Apologies again for the delay.
